### PR TITLE
Add ExternalRecipient tag

### PR DESCRIPTION
**Pull request by LunarX**

### CODE REVIEW

**Review by JorisBodin — APPROVED** [review]

File: app/src/main/java/com/infomaniak/mail/ui/MainViewModel.kt [hunk 1]
```diff
@@ -112,6 +112,10 @@ class MainViewModel @Inject constructor(
         it?.let(MailboxController::getMailbox)
     }.asLiveData(ioCoroutineContext)
 
+    val currentMailboxLive = _currentMailboxObjectId.flatMapLatest { objectId ->
+        objectId?.let(MailboxController::getMailboxAsync)?.mapNotNull { it.obj } ?: emptyFlow()
+    }.asLiveData(ioCoroutineContext)
+
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/ui/MainViewModel.kt:118]: This is used only in ThreadViewModel.

It should be moved to ThreadViewModel, and get the Mailbox via `AccountUtils.currentUserId` & `AccountUtils.currentMailboxId` instead of `_currentMailboxObjectId`.

**Review by KevinBoulongne — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/infomaniak/mail/ui/main/thread/RoundedBackgroundSpan.kt [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+        canvas.drawRoundRect(rect, CORNER_RADIUS, CORNER_RADIUS, paint)
+
+        paint.setGivenTextStyle()
+        canvas.drawText(
+            /* text = */ text,
+            /* start = */ start,
+            /* end = */ end,
+            /* x = */ LEFT_MARGIN + x + PADDING,
+            /* y = */ y.toFloat() - VERTICAL_OFFSET,
+            /* paint = */ paint
```
[the remark this answers is not in the document]
Comment (KevinBoulongne) [reply app/src/main/java/com/infomaniak/mail/ui/main/thread/RoundedBackgroundSpan.kt:76]: Nevertheless, don't forget the last comma.

File: app/src/main/java/com/infomaniak/mail/ui/main/newMessage/NewMessageFragment.kt [hunk 3]
```diff
@@ -256,7 +256,11 @@ class NewMessageFragment : Fragment() {
             autoComplete = autoCompleteTo,
             onAutoCompletionToggledCallback = { hasOpened -> toggleAutoCompletion(TO, hasOpened) },
             onContactAddedCallback = { newMessageViewModel.addRecipientToField(it, TO) },
-            onContactRemovedCallback = { newMessageViewModel.removeRecipientFromField(it, TO) },
+            onContactRemovedCallback = {
+                newMessageViewModel.removeRecipientFromField(it, TO)
+                updateBannerVisibility()
+
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/ui/main/newMessage/NewMessageFragment.kt:262]:
```suggestion
```

File: app/src/main/java/com/infomaniak/mail/ui/main/newMessage/RecipientFieldView.kt [hunk 4]
```diff
@@ -404,8 +416,37 @@ class RecipientFieldView @JvmOverloads constructor(
...
-        const val MAX_WIDTH_PERCENTAGE = 0.8
-        const val MAX_ALLOWED_RECIPIENT = 99
+    fun findAlreadyExistingExternalRecipientsInFields(): Pair<String?, Int> {
+        val recipients = contactChipAdapter.getRecipients().filter { it.displayAsExternal }
+        val recipientCount = recipients.count()
+        return (if (recipientCount == 1) recipients.single().email else null) to recipientCount
+    }
+
+    companion object {
+        private const val MAX_WIDTH_PERCENTAGE = 0.8
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/ui/main/newMessage/RecipientFieldView.kt:426]:
```suggestion
        private const val MAX_WIDTH_PERCENTAGE = 0.8f
```

File: app/src/main/java/com/infomaniak/mail/ui/main/thread/ThreadFragment.kt [hunk 5]
```diff
@@ -426,6 +445,97 @@ class ThreadFragment : Fragment() {
...
+        aliases: List<String>,
+        externalMailFlagEnabled: Boolean,
+    ): Pair<String, CharSequence> = with(binding) {
+        val subject = context.formatSubject(thread.subject)
+        if (!externalMailFlagEnabled) return subject to subject
+
+        val (externalRecipientEmail, externalRecipientQuantity) = UiUtils.findExternalRecipientsInThread(
+            thread,
+            emailDictionary,
+            aliases
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/ui/main/thread/ThreadFragment.kt:471]:
```suggestion
            aliases,
```

File: app/src/main/java/com/infomaniak/mail/ui/main/thread/ThreadFragment.kt [hunk 6]
```diff
@@ -426,6 +445,97 @@ class ThreadFragment : Fragment() {
...
+        setSpan(
+            RoundedBackgroundSpan(
+                backgroundColor = backgroundColor,
+                textColor = textColor,
+                textTypeface = textTypeface,
+                fontSize = textSize,
+            ),
+            startIndex,
+            endIndex,
+            Spannable.SPAN_EXCLUSIVE_EXCLUSIVE
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/ui/main/thread/ThreadFragment.kt:518]:
```suggestion
            Spannable.SPAN_EXCLUSIVE_EXCLUSIVE,
```

File: app/src/main/java/com/infomaniak/mail/ui/main/thread/ThreadFragment.kt [hunk 7]
```diff
@@ -426,6 +445,97 @@ class ThreadFragment : Fragment() {
...
+    ) {
+        // TODO : Currently, the clickable zone extends beyond the span up to the edge of the textview. This is the same
+        //  comportment that Gmail has. See if we can find a fix for this later
+        setSpan(
+            object : ClickableSpan() {
+                override fun onClick(widget: View) = onClick()
+            },
+            startIndex,
+            endIndex,
+            Spannable.SPAN_EXCLUSIVE_EXCLUSIVE
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/ui/main/thread/ThreadFragment.kt:535]:
```suggestion
            Spannable.SPAN_EXCLUSIVE_EXCLUSIVE,
```

**Review by KevinBoulongne — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/infomaniak/mail/utils/UiUtils.kt [hunk 8]
```diff
@@ -130,4 +133,61 @@ object UiUtils {
...
+    }
+
+    /**
+     * Only returns a quantity of at most 2, used to differentiate between the singular or plural form of the dialog messages
+     */
+    private fun findExternalRecipientInIterables(
+        emailDictionary: MergedContactDictionary,
+        aliases: List<String>,
+        vararg recipientLists: Iterable<Recipient>,
+    ): Pair<String?, Int> {
```
Comment (FabianDevel) [inline app/src/main/java/com/infomaniak/mail/utils/UiUtils.kt:175]: What about putting these 3 functions as Extension of Thread, Draft and Iterable<Recipient> ?

**Review by FabianDevel — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/infomaniak/mail/ui/main/newMessage/NewMessageActivity.kt [hunk 9]
```diff
@@ -63,10 +60,44 @@ class NewMessageActivity : BaseActivity() {
...
+        }
+
+        informationButton.setOnClickListener {
+            val description = resources.getQuantityString(
+                R.plurals.externalDialogDescriptionExpeditor,
+                externalRecipientQuantity,
+                externalRecipientEmail,
+            )
+
+            // TODO : Reuse instance
```
[the remark this answers is not in the document]
Comment (LunarX) [reply app/src/main/java/com/infomaniak/mail/ui/main/newMessage/NewMessageActivity.kt:86]: later, there's already a card

File: app/src/main/java/com/infomaniak/mail/ui/main/newMessage/NewMessageFragment.kt [hunk 10]
```diff
@@ -267,7 +271,10 @@ class NewMessageFragment : Fragment() {
             autoComplete = autoCompleteCc,
             onAutoCompletionToggledCallback = { hasOpened -> toggleAutoCompletion(CC, hasOpened) },
             onContactAddedCallback = { newMessageViewModel.addRecipientToField(it, CC) },
-            onContactRemovedCallback = { newMessageViewModel.removeRecipientFromField(it, CC) },
+            onContactRemovedCallback = {
+                newMessageViewModel.removeRecipientFromField(it, CC)
+                updateBannerVisibility()
+            },
```
[the remark this answers is not in the document]
Comment (LunarX) [reply app/src/main/java/com/infomaniak/mail/ui/main/newMessage/NewMessageFragment.kt:277]: If it's as little complex and closely defined space-wise like in here, I don't think we need to bother with factorizing something like that. It would be more bothersome than useful. Nonetheless, here, by factorizing it I could make it just a slightly bit more readable so actually why not, I factorized it anyway

File: app/src/main/java/com/infomaniak/mail/utils/UiUtils.kt (hunk 8 above)
Comment (LunarX) [reply app/src/main/java/com/infomaniak/mail/utils/UiUtils.kt:175]: Even better was what we talked about with Joris but ended up forgetting apparently. I extracted them to their own util

**Review by FabianDevel — CHANGES_REQUESTED** [review]

**Review by FabianDevel — APPROVED** [review]

File: app/src/main/java/com/infomaniak/mail/ui/main/thread/ThreadViewModel.kt [hunk 11]
```diff
@@ -70,6 +71,29 @@ class ThreadViewModel @Inject constructor(
         messageController.getSortedMessages(threadUid)?.asFlow()?.asLiveData()?.let { emitSource(it) }
     }
 
+    private val currentMailboxLive = MailboxController.getMailboxAsync(
+        AccountUtils.currentUserId,
+        AccountUtils.currentMailboxId,
+    ).map { it.obj }.asLiveData(ioCoroutineContext)
+
+    fun threadMergedContactAndMailboxMediator(
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/ui/main/thread/ThreadViewModel.kt:79]: The name is kinda cluttered.

Maybe `fun threadAndMergedContactAndMailboxMediator()` ?

File: app/src/main/java/com/infomaniak/mail/data/models/correspondent/Recipient.kt [hunk 12]
```diff
@@ -26,7 +28,11 @@ import kotlinx.serialization.Serializable
 
 @Parcelize
 @Serializable
-class Recipient : EmbeddedRealmObject, Correspondent {
+open class Recipient() : EmbeddedRealmObject, Correspondent {
+    constructor(email: String, name: String) : this() {
+        initLocalValues(email, name)
+    }
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/data/models/correspondent/Recipient.kt:34]: If we want to go this way, we should totally remove `initLocalValues()`, and add an explanatory comment to insist on the fact that we need to use this secondary constructor.

This is a big change, it should be done on another PR, and this diff should be reverted.

File: app/src/main/java/com/infomaniak/mail/ui/main/newMessage/NewMessageFragment.kt [hunk 13]
```diff
@@ -332,13 +333,29 @@ class NewMessageFragment : Fragment() {
         newMessageViewModel.isAutoCompletionOpened = isAutoCompletionOpened
     }
 
+    private fun Recipient.removeInViewModelAndUpdateBannerVisibility(type: FieldType) {
+        newMessageViewModel.removeRecipientFromField(this, type)
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/ui/main/newMessage/NewMessageFragment.kt:337]:
```suggestion
        newMessageViewModel.removeRecipientFromField(recipient = this, type)
```

File: app/src/main/java/com/infomaniak/mail/ui/main/newMessage/RecipientFieldView.kt [hunk 14]
```diff
@@ -404,8 +417,40 @@ class RecipientFieldView @JvmOverloads constructor(
...
+                chipStrokeWidth = NO_STROKE
+            }
+        }
+    }
+
+    @Parcelize
+    class StyledRecipient(
+        private val inputEmail: String,
+        private val inputName: String,
+        val displayAsExternal: Boolean
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/ui/main/newMessage/RecipientFieldView.kt:451]:
```suggestion
        val displayAsExternal: Boolean,
```

File: app/src/main/java/com/infomaniak/mail/ui/main/thread/ThreadFragment.kt [hunk 15]
```diff
@@ -426,6 +445,88 @@ class ThreadFragment : Fragment() {
...
+        externalMailFlagEnabled: Boolean,
+    ): Pair<String, CharSequence> = with(binding) {
+        val subject = context.formatSubject(thread.subject)
+        if (!externalMailFlagEnabled) return subject to subject
+
+        val (externalRecipientEmail, externalRecipientQuantity) = thread.findExternalRecipients(emailDictionary, aliases)
+        if (externalRecipientQuantity == 0) return subject to subject
+
+        val externalPostfix = getString(R.string.externalTag)
+        val postfixedSubject = "$subject$EXTERNAL_TAG_SEPARATOR$externalPostfix"
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/ui/main/thread/ThreadFragment.kt:472]:
```suggestion
        val postfixedSubject = "${subject}${EXTERNAL_TAG_SEPARATOR}${externalPostfix}"
```

File: app/src/main/java/com/infomaniak/mail/utils/Extensions.kt [hunk 16]
```diff
@@ -364,26 +364,45 @@ fun List<Message>.getFoldersIds(exception: String? = null) = mapNotNull { if (it
...
+fun Activity.createInformationDialog(
+    title: String,
+    description: CharSequence,
+    @StringRes confirmButtonText: Int,
+): AlertDialog = createDescriptionDialog(
+    title = title,
+    description = description,
+    confirmButtonText = confirmButtonText,
+    displayCancelButton = false,
+    onPositiveButtonClicked = {}
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/utils/Extensions.kt:382]:
```suggestion
    onPositiveButtonClicked = {},
```

**Review by KevinBoulongne — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/infomaniak/mail/data/models/correspondent/Recipient.kt (hunk 12 above)
Comment (LunarX) [reply app/src/main/java/com/infomaniak/mail/data/models/correspondent/Recipient.kt:34]: For a second, I believed I had no other option than using a constructor to do what I wanted and if I add the constructor I might as well use it everywhere. But actually it's not needed so yeah I can undo this
